be/src/olap/delete_bitmap_calculator.cpp

```diff
@@ -83,116 +91,209 @@ Status MergeIndexDeleteBitmapCalculatorContext::_next_batch(size_t row_id) {
+    auto lhs_seq = Slice(lhs.get_data() + lhs.get_size() - _sequence_length, _sequence_length);
+    auto rhs_seq = Slice(rhs.get_data() + rhs.get_size() - _sequence_length, _sequence_length);
+    return lhs_seq.compare(rhs_seq);
+}
+
+bool MergedPKIndexDeleteBitmapCalculatorContext::Comparator::is_key_same(Slice const& lhs,
+                                                                         Slice const& rhs) const {
+    return compare_key(lhs, rhs) == 0;
 }
 
-Status MergeIndexDeleteBitmapCalculator::init(RowsetId rowset_id,
-                                              std::vector<SegmentSharedPtr> const& segments,
-                                              size_t seq_col_length, size_t max_batch_size) {
-    _rowset_id = rowset_id;
+Status MergedPKIndexDeleteBitmapCalculator::init(std::vector<SegmentSharedPtr> const& segments,
+                                                 const std::vector<int64_t>* end_versions,
+                                                 size_t num_base_segments, size_t seq_col_length,
+                                                 size_t max_batch_size) {
     _seq_col_length = seq_col_length;
-    _comparator = MergeIndexDeleteBitmapCalculatorContext::Comparator(seq_col_length);
-    _contexts.reserve(segments.size());
+    _comparator = MergedPKIndexDeleteBitmapCalculatorContext::Comparator(seq_col_length);
+    size_t num_segments = segments.size();
+    _contexts.reserve(num_segments);
     _heap = std::make_unique<Heap>(_comparator);
 
-    for (auto& segment : segments) {
+    for (size_t i = 0; i < num_segments; ++i) {
+        auto&& segment = segments[i];
+        int64_t end_version = end_versions ? (*end_versions)[i] : 0;
+        segment->rowset_id();
         RETURN_IF_ERROR(segment->load_index());
         auto pk_idx = segment->get_primary_key_index();
         std::unique_ptr<segment_v2::IndexedColumnIterator> index;
         RETURN_IF_ERROR(pk_idx->new_iterator(&index));
         auto index_type = vectorized::DataTypeFactory::instance().create_data_type(
                 pk_idx->type_info()->type(), 1, 0);
-        _contexts.emplace_back(std::move(index), index_type, segment->id(), pk_idx->num_rows());
+        bool is_base_segment = i < num_base_segments;
+        _contexts.emplace_back(std::move(index), index_type, segment->rowset_id(), end_version,
+                               segment->id(), is_base_segment, pk_idx->num_rows(), max_batch_size);
         _heap->push(&_contexts.back());
     }
     return Status::OK();
 }
 
-Status MergeIndexDeleteBitmapCalculator::calculate_one(RowLocation& loc) {
-    // get the location of a out-of-date row
-    while (!_heap->empty()) {
-        auto cur_ctx = _heap->top();
+Status MergedPKIndexDeleteBitmapCalculator::init(
+        std::vector<SegmentSharedPtr> const& segments,
+        const std::vector<RowsetSharedPtr>* specified_rowsets,
+        bool calc_delete_bitmap_between_segments, size_t seq_col_length, size_t max_batch_size) {
```

Comment:
parameter `calc_delete_bitmap_between_segments` seems not necessary, you can use `specified_rowsets == nullptr` instead?